Add contributor guidelines

s-ben opened the pull request:
Fleshed out the Contributor Guidelines page. Per this [chat](https://matrix.to/#/!tfqymymiNgzSUJTHqS:decred.org/$156797114324642xSCzC:decred.org?via=decred.org&via=matrix.org&via=zettaport.com), this is basically porting over the developer-centric content from the guidelines in dcrdocs and adding some general high-level guidance. 

I did a little polishing on the [GitHub instructions](https://docs.decred.org/contributing/using-github/) from dcrdocs. They're looking pretty good now, but could definitely use some review. This could be a good opportunity for devs to weigh in and make corrections or improvements.

jholdstock commented on docs/contributing/contributor-guidelines.md:
```diff
@@ -1,3 +1,31 @@
+
+## Finding Work
+
+Nearly all work occurs in public repos and is open to new contributors. A full list of repos can be found in the Decred [GitHub org](https://github.com/decred). A list of projects under active development and descriptions of them can be found on the [Projects](../../projects) page. 
```
Let's spells out "organisation" rather than "org". It may not be obvious for those new to github.

Also the link needs to be to the `.md` file - I'm sure one day you will learn :laughing:

s-ben replied:
Done.

>Also the link needs to be to the .md file - I'm sure one day you will learn 😆

It's a bad habbit. I can quit anytime :man_facepalming:

jholdstock commented on docs/contributing/contributor-guidelines.md:
```diff
@@ -1,3 +1,31 @@
+
+## Contributing
+
+Once you've found an issue you'd like to work on, announce your plans **before** starting work. This avoids duplicate work, minimizes the chances you're working on something that might not fit with the consensus of the community, and permits discussions that can help you achieve your goal. Most projects have a dedicated chat channel for discussing project-specific work. 
```
We should explain **how** to do the announcement. Typically commenting on the issue or by speaking with relevant devs on Matrix.

s-ben replied:
Agreed. Reworded to:

>Once you've found something you'd like to work on, announce your plans **before** starting work.  This avoids duplicate work, minimizes the chances you're working on something that might not fit with the consensus of the community, and permits discussions that can help you achieve your goal. If working on a GitHub Issue, you can do this by commenting on the Issue. You can also speak with the relevent devs on Matrix. Most projects have a dedicated chat channel for discussing project-specific work.

jholdstock commented on docs/contributing/contributor-guidelines.md:
```diff
@@ -1,3 +1,31 @@
+
+## GitHub
+
+Whie different repos may have slightly different GitHub processes, below are the basic steps common to all repos.
```
While*

s-ben replied:
Fixed

jholdstock commented on docs/contributing/contributor-guidelines.md:
```diff
@@ -1,3 +1,31 @@
+1. Clone the forked repo. 
+1. Make your changes in a local branch.
+1. When your changes are ready to be reviewed (or you just want input from other devs), push your branch to your forked repo and open a Pull Request (PR) on GitHub against the main repo. If your changes address an Issue, put `Closes #` or `Fixes #`in the PR description. This will allow GitHub to automatically close the Issue when the PR is merged. 
+1. All changes must be reviewed and receive at least one approval before they can be merged. Only team members can give official approval, but comments from other users are encouraged.
```
To be specific, "Only team members **with write access on the repository** can give official approval"

s-ben replied:
Thought that's what it should be. Using this wording now

jholdstock commented on docs/contributing/contributor-guidelines.md:
```diff
@@ -1,3 +1,31 @@
+1. All changes must be reviewed and receive at least one approval before they can be merged. Only team members can give official approval, but comments from other users are encouraged.
+1. If a reviewer requests changes, make those changes in the local branch you were working in. Then push those changes to your forked repo. This will automatically update the PR in a way that allows the reviewer to see the changes you've made.
+1. Once approved, your PR can be merged into master. To keep history clean, only non-fast-forward merges are allowed (i.e. we want a linear history). If a change consists of multiple commits, those commits typically must be squashed into a single commit. If there is a reason to include multiple commits in a PR, that will be considered on a case-by-case basis. 
+1. If your PR consists of a single commit (or multiple commits that can be automatically squased by GitHub), and is caught up with master, the reviewer will merge your PR. If your branch is too far behind master, you may be asked to rebase your commit and push again. The reviewer will then merge your PR. 
```
squashed*

jholdstock replied:
"is caught up with master" => "has no conflicts with master"

jholdstock replied:
I think the final sentence "The reviewer will then merge your PR." is unnecessary/repetitive. Maybe phrase differently or remove?

s-ben replied:
Accepted all of these fixes/suggestions

s-ben commented:
@jholdstock addressed comments, accepted all suggestions. Glad to see the corrections on the GitHub process. Was thinking some of the instructions from dcrdocs were a bit off but wasn't sure.
